arwen/contexts/metering.go

```diff
@@ -99,20 +105,204 @@ func (context *meteringContext) PopDiscard() {
 	context.stateStack = context.stateStack[:stateStackLen-1]
 }
 
+func (context *meteringContext) PopMergeActiveState() {
+	stateStackLen := len(context.stateStack)
+	if stateStackLen == 0 {
+		return
+	}
+
+	prevState := context.stateStack[stateStackLen-1]
+	context.stateStack = context.stateStack[:stateStackLen-1]
+
+	context.initialGasProvided = prevState.initialGasProvided
+	context.initialCost = prevState.initialCost
+	context.gasForExecution = prevState.gasForExecution
+
+	context.addToGasUsedByAccounts(prevState.gasUsedByAccounts)
+}
+
+func (context *meteringContext) cloneGasUsedByAccounts() map[string]uint64 {
+	clone := make(map[string]uint64, len(context.gasUsedByAccounts))
+
+	for address, gasUsed := range context.gasUsedByAccounts {
+		clone[address] = gasUsed
+	}
+
+	return clone
+}
+
+func (context *meteringContext) addToGasUsedByAccounts(gasUsed map[string]uint64) {
+	for address, gas := range gasUsed {
+		context.gasUsedByAccounts[address] += gas
+	}
+}
+
+func (context *meteringContext) UpdateGasStateOnSuccess(vmOutput *vmcommon.VMOutput) error {
+	context.updateSCGasUsed()
+	context.setGasUsedToOutputAccounts(vmOutput)
+
+	err := context.checkGasLegacy()
```

Comment:
`context.checkGasLegacy()` is a no-op if `host.IsArwenV2Enabled()`. It's only here for backwards compatibility, because the actual `context.checkGas()` will report errors for older transactions, before gas forwarding and the gas invariant were introduced.

Reply:
we do not care about legacy as we will resolve this from the factory.

Reply:
Deleted.